crates/runtime/src/tool_use.rs

```diff
@@ -0,0 +1,349 @@
+You may obtain a copy of the License at
+https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#![allow(clippy::missing_errors_doc)]
+use std::pin::Pin;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use arrow::array::RecordBatch;
+use itertools::Itertools;
+use llms::chat::{Chat, Result as ChatResult};
+
+use async_openai::error::OpenAIError;
+use async_openai::types::{
+    ChatCompletionMessageToolCall, ChatCompletionRequestAssistantMessageArgs,
+    ChatCompletionRequestMessage, ChatCompletionRequestToolMessageArgs,
+    ChatCompletionResponseStream, ChatCompletionTool, ChatCompletionToolChoiceOption,
+    ChatCompletionToolType, CreateChatCompletionRequest, CreateChatCompletionResponse,
+    FunctionObject,
+};
+
+use async_openai::types::CreateChatCompletionRequestArgs;
+use async_trait::async_trait;
+use futures::Stream;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use snafu::ResultExt;
+
+use crate::Runtime;
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum SpiceToolsOptions {
+    Auto,
+    Disabled,
+    Specific(Vec<String>),
+}
+
+impl SpiceToolsOptions {
+    // Check if spice tools can be used.
+    pub fn can_use_tools(&self) -> bool {
+        match self {
+            SpiceToolsOptions::Auto => true,
+            SpiceToolsOptions::Disabled => false,
+            SpiceToolsOptions::Specific(t) => !t.is_empty(),
+        }
+    }
+
+    /// Filter out a list of tools permitted by the  [`SpiceToolsOptions`].
+    pub fn filter_tools(&self, tools: Vec<Box<dyn SpicedTool>>) -> Vec<Box<dyn SpicedTool>> {
+        match self {
+            SpiceToolsOptions::Auto => tools,
+            SpiceToolsOptions::Disabled => vec![],
+            SpiceToolsOptions::Specific(t) => tools
+                .into_iter()
+                .filter(|tool| t.contains(&tool.name()))
+                .collect(),
+        }
+    }
+}
+
+impl FromStr for SpiceToolsOptions {
+    type Err = Box<dyn std::error::Error + Send + Sync>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.trim().to_lowercase().as_str() {
+            "auto" => Ok(SpiceToolsOptions::Auto),
+            "disabled" => Ok(SpiceToolsOptions::Disabled),
+            _ => Ok(SpiceToolsOptions::Specific(
+                s.split(',')
+                    .map(|item| item.trim().to_string())
+                    .filter(|item| !item.is_empty())
+                    .collect(),
+            )),
+        }
+    }
+}
+
+/// Tools that implement this trait can be automatically used by LLMs in the runtime.
+#[async_trait]
+pub trait SpicedTool: Sync + Send {
```

Comment:
```suggestion
pub trait SpiceModelTool: Sync + Send {
```

What do you think about this name